_posts/2022-01-21-skyhook-bringing-computation-to-storage-with-apache-arrow.md

```diff
@@ -0,0 +1,219 @@
+
+Then, Skyhook defines a custom “file format” in the Arrow Datasets layer.
+Queries against such files get translated to direct requests to Ceph using those new operations, bypassing the traditional POSIX file system layer.
+After decoding, filtering, and projection, Ceph directly sends the client Arrow record batches, minimizing CPU overhead for encoding/decoding—another optimization Arrow makes possible.
+The record batches use Arrow’s compression support to further save bandwidth.
+
+<figure>
+  <img src="{{ site.baseurl }}/img/20220121-skyhook-architecture.png"
+       alt="Skyhook Architecture"
+       width="100%" class="img-responsive">
+  <figcaption>
+    Skyhook extends Ceph and Arrow Datasets to push queries down to Ceph, reducing the client workload and network traffic.
+  </figcaption>
+</figure>
+
+Skyhook also optimizes how Parquet files in particular are stored.
+Parquet files consist of a series of row groups, which each contain a chunk of the rows in a file.
+When storing such files, Skyhook either pads or splits them so that each row group is stored as its own Ceph object.
+By striping or splitting the file in this way, we can parallelize scanning at sub-file granularity across the Ceph nodes for further performance improvements.
+
+## Applications
+
+In benchmarks, Skyhook has minimal storage-side CPU overhead and virtually eliminates client-side CPU usage.
+And scaling the storage cluster decreases query latency commensurately.
+For systems like Dask that use the Arrow Datasets API, this means that just by switching to the Skyhook file format, we can speed up dataset scans, reduce the amount of data that needs to be transferred, and free up CPU resources for computations.
+
+<!-- TODO: add the updated figure from Jayjeet -->
+<figure>
+  <img src="{{ site.baseurl }}/img/20220121-skyhook-architecture.png"
+       alt="In benchmarks, Skyhook reduces client CPU usage while minimally impacting storage cluster CPU usage."
+       width="100%" class="img-responsive">
+  <figcaption>
+    Skyhook frees the client CPU to do useful work, while minimally impacting the work done by the storage machines. Figure sourced from TODO.
+  </figcaption>
+</figure>
+
+Of course, the ideas behind Skyhook apply to other systems adjacent to and beyond Apache Arrow.
+For example, “lakehouse” systems like Apache Iceberg and Delta Lake also build on distributed storage systems, and can naturally benefit from Skyhook to offload computation.
+Additionally, in-memory SQL-based query engines like [DuckDB][duckdb], which integrate seamlessly with Apache Arrow, can benefit from Skyhook by offloading portions of SQL queries.
+
+## Summary and Acknowledgements
+
+<!-- TODO: is there documentation on how to deploy the plugin? Can't find anything on ceph.io -->
```

Comment:
Does anyone know where on the Ceph website this is documented? Worst case we could link to the integration_skyhook.sh script but I'd prefer to link some docs on how to deploy this.

Reply:
@carlosmalt do you know of some official documentation regarding this ? Unfortunately, I am not aware of any.

Reply:
I see instructions in your blog post, so for now, I'll link there.